sljit_src/sljitExecAllocator.c

```diff
@@ -91,78 +91,88 @@ static SLJIT_INLINE void free_chunk(void *chunk, sljit_uw size)
 	VirtualFree(chunk, 0, MEM_RELEASE);
 }
 
-#else
+#else /* POSIX */
 
 #ifdef __APPLE__
-#ifdef MAP_ANON
-/* Configures TARGET_OS_OSX when appropriate */
-#include <TargetConditionals.h>
-
-#if TARGET_OS_OSX && defined(MAP_JIT)
-#include <sys/utsname.h>
-#endif /* TARGET_OS_OSX && MAP_JIT */
-
-#ifdef MAP_JIT
-
+#include <AvailabilityMacros.h>
 /*
    On macOS systems, returns MAP_JIT if it is defined _and_ we're running on a
-   version where it's OK to have more than one JIT block.
+   version where it's OK to have more than one JIT block or where MAP_JIT is
+   required.
    On non-macOS systems, returns MAP_JIT if it is defined.
 */
+#ifdef MAP_JIT
+#include <TargetConditionals.h>
+#if TARGET_OS_OSX && (defined SLJIT_CONFIG_X86 && SLJIT_CONFIG_X86)
+#ifdef MAP_ANON
+#include <sys/utsname.h>
+#include <stdlib.h>
+
+#define SLJIT_MAP_JIT	(get_map_jit_flag())
+
 static SLJIT_INLINE int get_map_jit_flag()
 {
-#if TARGET_OS_OSX
-	sljit_sw page_size = get_page_alignment() + 1;
+	sljit_sw page_size;
 	void *ptr;
+	struct utsname name;
+	int os_kernel_major;
 	static int map_jit_flag = -1;
 
-	/*
-	  The following code is thread safe because multiple initialization
-	  sets map_jit_flag to the same value and the code has no side-effects.
-	  Changing the kernel version witout system restart is (very) unlikely.
-	*/
-	if (map_jit_flag == -1) {
-		struct utsname name;
-
+	if (map_jit_flag < 0) {
 		map_jit_flag = 0;
 		uname(&name);
-
-		/* Kernel version for 10.14.0 (Mojave) */
-		if (atoi(name.release) >= 18) {
+		os_kernel_major = atoi(name.release);
+
+		/* this is likely to cause problems as seen in PCRE BUG2334 */
```

Comment:
FWIW I haven't seen any of the reported problems while testing in 10.15.1 (Catalina), so it might be as well safe to use MAP_JIT by default.

obviously it will cause mmap to fail if the hardened runtime is in use AND the application doesn't have the com.apple.security.cs.allow-jit entitlement, but that is the expected behaviour.

MAP_JIT doesn't prevent creating more than 1 map or inheriting the map through fork() like it does in 10.13.6